lib/jsonapi/serializer.rb

```diff
@@ -1,12 +1,73 @@
-require 'fast_jsonapi'
+require 'jsonapi/serializer/base'
+require 'jsonapi/serializer/core'
+require 'jsonapi/serializer/dsl'
+require 'jsonapi/serializer/errors'
+require 'jsonapi/serializer/trackable'
 
 module JSONAPI
   module Serializer
-    # TODO: Move and cleanup the old implementation...
+    extend Trackable
+
+    # Self registers any inherited/extended class to keep track of it
+    #
+    # @return nothing
     def self.included(base)
+      super
+
+      register_serializer(base)
+
       base.class_eval do
-        include FastJsonapi::ObjectSerializer
+        include ::JSONAPI::Serializer::Base
+        extend ::JSONAPI::Serializer::DSL
+        extend ::JSONAPI::Serializer::Core
+      end
+    end
+
+    # Serializes an object or a collection
+    #
+    # The `options` are passed to the serializer instance and follow the
+    # same purpose.
+    #
+    # The only extra option is the `options[:serializers]`
+    # one can use to pass a mapping of object to serializer classes to
+    # be strict about how each object is serialized.
+    #
+    # @param object_or_collection [Object] to be serialized
+    # @param options [Hash] serialization parameters
+    # @return [Hash] of serialized JSONAPI data
+    def self.serialize(object_or_collection, options: nil)
+      options ||= {}
+      options[:is_collection] = is_collection?(
```

Comment:
it's dangerous to modify original hash. Imagine I want to use same options on for multiple serializations:
```ruby
my_options = {}
JSONAPI::Serializer.serialize(some_collection, my_options)
my_options # =>  { is _collection: true }
JSONAPI::Serializer.serialize(some_record, my_options) # this will be treated as collection because options was modified :(
```

Reply:
Good call @povilasjurcys, I'll make sure we're good here! Thank you!